text/0000-programmatic-style-injection.md

```diff
@@ -0,0 +1,172 @@
+
+The `stylesheets` property works on both light DOM and shadow DOM LWC components. This stylesheet injection mechanism complements the existing implicit template/stylesheet resolution mechanism.
+
+The LWC engine caches the static `stylesheets` value for the lifetime of the application during the component class definition. Modifying the `stylesheets` property after the component code is evaluated will not affect the stylesheet injected by this component. 
+
+For component inheritance, the LWC engine doesn't attempt to merge ancestor components stylesheets. Developers are in charge of merging (or not) the ancestors' stylesheets with their stylesheets.
+
+```js
+import { LightningElement } from 'lwc';
+import ancestorStylesheet from './ancestor.css';
+import childStylesheet from './child.css';
+
+class Ancestor extends LightningElement {
+    static stylesheets = [
+        ancestorStylesheet
+    ];
+}
+
+class Descendent extends Ancestor {
+    static stylesheets = [
+        ...super.stylesheets, 
+        childStylesheet
+    ];
+}
+```
+
+### Injection order
+
+The LWC engine injects stylesheets in the following order: first, the implicit stylesheet associated with the template and then all the stylesheets associated with the `stylesheets` property on the component class. This order allows the styles that are programmatically associated with the component to override the styles that are implicitly loaded via the template.
+
+In the following example the `<x-foo>` stylesheets will be injected in the following order: `foo.css`, `custom-a.css` and `custom-b.css`.
+
+```
+x/
+└── foo/
+    ├── custom-a.css
+    ├── custom-b.css
+    ├── foo.js
+    ├── foo.html
+    └── foo.css
+```
+
+```js
+// x/foo.js
+import { LightningElement } from 'lwc';
+import customStyleA from './custom-a.css';
+import customStyleB from './custom-b.css';
+
+export default Foo class extends LightningElement {
+    static stylesheets = [
+        customStyleA,
+        customStyleB
+    ];
+}
+```
+
+### LWC stylesheets modules
+
+Programmatically importing a stylesheet module from JavaScript using an import declaration is not a new concept introduced by this RFC. Such import is already supported by the LWC compiler. In fact, the template compiler is already using this mechanism to import the stylesheet implicitly associated with the template.
+
+However, until now, the compile-time behavior and runtime behavior for stylesheet modules have never been fully specified. This section fills this specification gap.
+
+Stylesheets modules are identifiable by their `.css` extension. This kind of module only exposes a single default export. The exported value is an opaque object, in other words, the internal implementation of the stylesheet is not exposed to developers. Developers shouldn't rely on the shape of those objects as it may change.

```

Comment:
@pmdartus After discussion with @rui-rayqiu, I'm not sure we need to make the stylesheet objects opaque. Instead, I would propose we use a similar system to the one we currently use for templates (`isTemplateRegistered`).

In short: there is not much benefit to making the object truly opaque. Sure, developers will see something like `LWCCSSStyleSheet` in the console rather than a function. But we don't do this for templates, so why bother for stylesheets? At the end of the day, we just want to disallow people from using arbitrary string-returning-functions, and we can do this with a `registerStylesheet()` utility.

We already [experimented with this](https://github.com/salesforce/lwc/pull/2708), but we didn't do it out of fear of breakages. If we could do it in a non-observable way (e.g. warn rather than throw an error), then I think this is feasible.

Reply:
> At the end of the day, we just want to disallow people from using arbitrary string-returning-functions, and we can do this with a `registerStylesheet()` utility.

My idea was to future-proof the stylesheet function by converting it into an object and at the same time making it opaque. By turning it into an object we could do a brand-check via the `instanceof` operator instead of storing this into a `WeakMap` and expose new methods on this object in the future if necessary. 

At the end of the day, in JavaScript, a function is just a special kind of object. Is it a deal breaker? I don't think so. We can certainly keep what we already have in place today. 

I don't have a very strong opinion on this specific aspect of the proposal. If you feel migration overhead and the potential cost of dealing with breakage outweigh the benefits, I am fine with stripping this from the proposal.

Reply:
@rui-rayqiu @nolanlawson What approach would you like to pursue?

Reply:
1. Add reporting to catch when someone is mutating a stylesheet (either setting `stylesheet.$scoped$` or using their own function-that-returns-a-string)
2. Mark this behavior as deprecated (e.g. in console warnings)
3. Roll out a "registered" and `Object.freeze`'d stylesheet function which disallows mutations

We already have a similar plan for the "frozen template": https://github.com/salesforce/lwc/issues/2826